eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala

```diff
@@ -75,7 +75,8 @@ class Switchboard(nodeParams: NodeParams, watcher: ActorRef, relayer: ActorRef,
       // if this is an incoming connection, we might not yet have created the peer
       val peer = createOrGetPeer(authenticated.remoteNodeId, offlineChannels = Set.empty)
       val features = nodeParams.featuresFor(authenticated.remoteNodeId)
-      val doSync = nodeParams.syncWhitelist.isEmpty || nodeParams.syncWhitelist.contains(authenticated.remoteNodeId)
+      // if the peer is whitelisted, we sync with them, otherwise we only sync with peers with whom we have at least one channel
+      val doSync = nodeParams.syncWhitelist.contains(authenticated.remoteNodeId) || (nodeParams.syncWhitelist.isEmpty && peersWithChannels.contains(authenticated.remoteNodeId))
```

Comment:
No, that's why we need what's done in the second commit, to trigger sync when a new channel is created.

Reply:
It seems that nodes we don't have channels with anymore will be kept in the  `peersWithChannels` map until the next restart and we will still sync with them

Reply:
I think it's not an issue, we won't because once we don't have channels anymore we disconnect and forget the peer

Reply:
Yes but if that peer connects again we will sync with them even if they're not white-listed and don't have a channel with us ?

Reply:
Good point, I'll see if we can easily fix that scenario.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1587/commits/2b5c07959a563076512994cf7c614d3a45456110, let me know what you think.

Reply:
A more immutable/functional version:

```diff
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala b/eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala
index 2d46e4a84..b9d048b56 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala
@@ -25,8 +25,6 @@ import fr.acinq.eclair.channel._
 import fr.acinq.eclair.remote.EclairInternalsSerializer.RemoteTypes
 import fr.acinq.eclair.router.Router.RouterConf

-import scala.collection.mutable
-
 /**
  * Ties network connections to peers.
  * Created by PM on 14/02/2017.
@@ -39,7 +37,7 @@ class Switchboard(nodeParams: NodeParams, watcher: ActorRef, relayer: ActorRef,
   context.system.eventStream.subscribe(self, classOf[PeerLastChannelClosed])

   // we load channels from database
-  val peersWithChannels: mutable.Set[PublicKey] = {
+  private def initialPeersWithChannels: Set[PublicKey] = {
     // Check if channels that are still in CLOSING state have actually been closed. This can happen when the app is stopped
     // just after a channel state has transitioned to CLOSED and before it has effectively been removed.
     // Closed channels will be removed, other channels will be restored.
@@ -51,10 +49,12 @@ class Switchboard(nodeParams: NodeParams, watcher: ActorRef, relayer: ActorRef,

     val peerChannels = channels.groupBy(_.commitments.remoteParams.nodeId)
     peerChannels.foreach { case (remoteNodeId, states) => createOrGetPeer(remoteNodeId, offlineChannels = states.toSet) }
-    mutable.Set.empty.addAll(peerChannels.keys)
+    peerChannels.keySet
   }

-  def receive: Receive = {
+  def receive: Receive = normal(initialPeersWithChannels)
+
+  def normal(peersWithChannels: Set[PublicKey]): Receive = {

     case Peer.Connect(publicKey, _) if publicKey == nodeParams.nodeId =>
       sender ! Status.Failure(new RuntimeException("cannot open connection with oneself"))
@@ -84,9 +84,9 @@ class Switchboard(nodeParams: NodeParams, watcher: ActorRef, relayer: ActorRef,
       val doSync = nodeParams.syncWhitelist.contains(authenticated.remoteNodeId) || (nodeParams.syncWhitelist.isEmpty && peersWithChannels.contains(authenticated.remoteNodeId))
       authenticated.peerConnection ! PeerConnection.InitializeConnection(peer, nodeParams.chainHash, features, doSync)

-    case ChannelIdAssigned(_, remoteNodeId, _, _) => peersWithChannels.add(remoteNodeId)
+    case ChannelIdAssigned(_, remoteNodeId, _, _) => context become normal(peersWithChannels + remoteNodeId)

-    case PeerLastChannelClosed(_, remoteNodeId) => peersWithChannels.remove(remoteNodeId)
+    case PeerLastChannelClosed(_, remoteNodeId) => context become normal(peersWithChannels - remoteNodeId)

     case Symbol("peers") => sender ! context.children

```

Reply:
Done in https://github.com/ACINQ/eclair/pull/1587/commits/c87bbe4c2f284ed6cecbb5bdcf289bde5d4a9e7a